diff -Nrup xinit-1.2.0/xinit.c xvt/xinit.c
--- xinit-1.2.0/xinit.c	2009-10-13 12:34:40.000000000 -0300
+++ xvt/xinit.c	2009-11-24 10:41:23.000000000 -0200
@@ -144,7 +144,7 @@ char xserverrcbuf[256];
 
 static char *default_server = "X";
 static char *default_display = ":0";        /* choose most efficient */
-static char *default_client[] = {"xterm", "-geometry", "+1+1", "-n", "login", NULL};
+static char *default_client[] = {"xvt", "-geometry", "+1+1", NULL};
 static char *serverargv[100];
 static char *clientargv[100];
 static char **server = serverargv + 2;        /* make sure room for sh .xserverrc args */
diff -Nrup xinit-1.2.0/man/xinit.man xvt/man/xinit.man
--- xinit-1.2.0/man/xinit.man	2009-11-05 17:07:28.000000000 -0200
+++ xvt/man/xinit.man	2009-11-24 10:43:27.000000000 -0200
@@ -54,7 +54,7 @@ If no specific client program is given o
 called \fI.xinitrc\fP to run as a shell script to start up client programs.
 If no such file exists, \fBxinit\fP will use the following as a default:
 .sp
-	xterm  \-geometry  +1+1  \-n  login  \-display  :0
+	xvt  \-geometry  +1+1  \-display  :0
 .sp
 .PP
 If no specific server program is given on the command line,
