diff -up xinit-1.4.4/startx.in.3~ xinit-1.4.4/startx.in
--- xinit-1.4.4/startx.in.3~	2025-03-11 18:23:31.243628440 +0100
+++ xinit-1.4.4/startx.in	2025-03-11 18:30:47.439812110 +0100
@@ -64,8 +64,6 @@ sysclientrc="@XINITDIR@/xinitrc"
 userserverrc="$HOME/.xserverrc"
 [ -f "${XSERVERRC}" ] && userserverrc="${XSERVERRC}"
 sysserverrc="$xinitdir/xserverrc"
-defaultclient="$xterm"
-defaultserver="$xserver"
 defaultclientargs=""
 defaultserverargs=""
 defaultdisplay=""
@@ -151,6 +149,12 @@ done
 defaultdisplay=":$d"
 unset d
 
+if [ -f $userclientrc ]; then
+    defaultclientargs=$userclientrc
+elif [ -f $sysclientrc ]; then
+    defaultclientargs=$sysclientrc
+fi
+
 whoseargs="client"
 while [ "$1" != "" ]; do
     case "$1" in
@@ -158,13 +162,13 @@ while [ "$1" != "" ]; do
     /''*|\./''*)
 	if [ "$whoseargs" = "client" ]; then
 	    if [ "$client" = "" ] && [ "$clientargs" = "" ]; then
-		client="$1"
+		clientargs="$1"
 	    else
 		clientargs="$clientargs $1"
 	    fi
 	else
 	    if [ "$server" = "" ] && [ "$serverargs" = "" ]; then
-		server="$1"
+		serverargs="$1"
 	    else
 		serverargs="$serverargs $1"
 	    fi
@@ -175,12 +179,17 @@ while [ "$1" != "" ]; do
 	;;
     *)
 	if [ "$whoseargs" = "client" ]; then
-	    clientargs="$clientargs $1"
+	    if [ "x$clientargs" = x ]; then
+	        clientargs="$defaultclientargs $1"
+	    else
+	        clientargs="$clientargs $1"
+	    fi
 	else
 	    # display must be the FIRST server argument
 	    if [ "$serverargs" = "" ] && \
 		 expr "$1" : ':[0-9][0-9]*$' > /dev/null 2>&1; then
 		display="$1"
+		serverargs="$1"
 	    else
 		serverargs="$serverargs $1"
 	    fi
@@ -190,29 +199,12 @@ while [ "$1" != "" ]; do
     shift
 done
 
-# process client arguments
-if [ "$client" = "" ]; then
-    client="$defaultclient"
-
-    # For compatibility reasons, only use startxrc if there were no client command line arguments
-    if [ "$clientargs" = "" ]; then
-        if [ -f "$userclientrc" ]; then
-            client="$userclientrc"
-        elif [ -f "$sysclientrc" ]; then
-            client="$sysclientrc"
-        fi
-    fi
-fi
-
 # if no client arguments, use defaults
 if [ "$clientargs" = "" ]; then
     clientargs="$defaultclientargs"
 fi
 
-# process server arguments
 if [ "$server" = "" ]; then
-    server="$defaultserver"
-
 if [ "$(uname -s)" = "Linux" ] ; then
     # When starting the defaultserver start X on the current tty to avoid
     # the startx session being seen as inactive:
@@ -223,15 +215,6 @@ if [ "$(uname -s)" = "Linux" ] ; then
         vtarg="vt$tty_num -keeptty"
     fi
 fi
-
-    # For compatibility reasons, only use xserverrc if there were no server command line arguments
-    if [ "$serverargs" = "" ] && [ "$display" = "" ]; then
-	if [ -f "$userserverrc" ]; then
-	    server="$userserverrc"
-	elif [ -f "$sysserverrc" ]; then
-	    server="$sysserverrc"
-	fi
-    fi
 fi
 
 # if no server arguments, use defaults
@@ -319,16 +302,8 @@ EOF
     done
 fi
 
-case "$(uname -s)" in
-CYGWIN_NT*|Darwin)
-    eval "$xinit" \"$client\" $clientargs -- \"$server\" $display $serverargs
-    ;;
-*)
-    "$xinit" "$client" $clientargs -- "$server" $display $serverargs
-    ;;
-esac
-retval=$?
-
+cleanup() {
+[ -n "$PID" ] && kill $PID > /dev/null 2>&1
 if [ "$enable_xauth" = 1 ] ; then
     if [ "$removelist" != "" ]; then
         "$xauth" remove $removelist
@@ -348,5 +323,14 @@ fi
 if [ "$(uname -s)" = "SunOS" ]; then
     kbd_mode -a
 fi
+}
+
+trap cleanup 0
+
+xinit $clientargs -- $serverargs -deferglyphs 16 &
+PID=$!
+wait $PID
+retval=$?
+unset PID
 
 exit $retval
