*** xinit-1.3.4/xinit.c.orig	2014-03-25 10:20:26.000000000 +0100
--- xinit-1.3.4/xinit.c	2014-10-10 00:55:14.000000000 +0200
***************
*** 560,565 ****
--- 560,566 ----
  {
      clientpid = fork();
      if (clientpid == 0) {
+         int fd;
          set_environment();
          setWindowPath();
  
***************
*** 567,573 ****
              Error("cannot change uid");
              _exit(EXIT_FAILURE);
          }
!         setpgid(0, getpid());
          Execute(client_argv);
          Error("Unable to run program \"%s\"", client_argv[0]);
  
--- 568,585 ----
              Error("cannot change uid");
              _exit(EXIT_FAILURE);
          }
! 
!         fd = open ("/dev/null", O_RDONLY);
! 
!         if (fd < 0) {
!                 Error("cannot open /dev/null: %s\n", strerror(errno));
!                 _exit(EXIT_FAILURE);
!         }
!         close (STDIN_FILENO);
!         dup2 (fd, STDIN_FILENO);
!         close (fd);
!         setsid();
! 
          Execute(client_argv);
          Error("Unable to run program \"%s\"", client_argv[0]);
  
